BJS2-65863-write-recommendation

marsel-mkh opened the pull request.

Commit:
BJS2-65863-write-recommendation

Zhltsk-V commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,238 @@
+        LocalDateTime sixMonthsAgo = LocalDateTime.now().minusMonths(6);
+        LocalDateTime recommendationDate = recommendation.getUpdatedAt();
+
+        if (recommendationDate.isAfter(sixMonthsAgo)) {
```
Сейчас код проверяет, что дата обновления более свежая, чем 6 месяцев назад, что не соответствует требованию.
Необходимо использовать - isBefore(sixMonthsAgo).

Zhltsk-V commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,238 @@
+        String content = recommendation.getContent();
+        recommendationRepository.update(authorId, receiverId, content);
+
+        skillOfferRepository.deleteAllByRecommendationId(recommendation.getId());
```
Удаление всех SkillOffer перед обновлением рекомендации может привести к лишним операциям. Лучше удалять только те SkillOffer, которые не входят в обновленный список.

Можно сравнить старые и новые SkillOffer и удалить только те, которые больше не нужны.

Zhltsk-V commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,57 @@
+        return recommendationService.update(updated);
+    }
+
+    private void deleteRecommendation(long recommendationId) {
```
Точно ли тут должен быть модификатор доступа private?

Zhltsk-V commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+
+    private void validationRecommendation(RecommendationDto recommendation) {
```
Можно проверить, что authorId и receiverId не равны (пользователь не может рекомендовать сам себе).

Zhltsk-V commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,238 @@
+     * @param receiverId - идентификатор пользователя
+     * @return List<RecommendationDto> - список полученных рекомендаций
+     */
+    public List<RecommendationDto> getAllUserRecommendations(long receiverId) {
```
Дублирование кода в методах getAllUserRecommendations и getAllGivenRecommendations. Можно вынести общую логику в отдельный приватный метод.

Zhltsk-V commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,238 @@
+     * @return List<RecommendationDto> - список полученных рекомендаций
+     */
+    public List<RecommendationDto> getAllUserRecommendations(long receiverId) {
+        Pageable pageable = PageRequest.of(0, Integer.MAX_VALUE);
```
Pageable с фиксированным размером может привести к проблемам с производительностью, если рекомендаций будет очень много.

marsel-mkh replied:
Насчет этого нужно разобраться, в репезитории нет реализации указанной в jire

RinatBeybutov replied:
обычно Pageable приходит с фронта и нам ничего не нужно выдумывать)

RinatBeybutov replied:
изменений не вижу

Commit:
BJS2-65863-write-recommendation
FIX: mapper, add SkillOfferMapper, fix validateRecommendation

RinatBeybutov commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationDto.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+
+@Data
+public class RecommendationDto {
```
Обсуждали правило по именованию дто: если на создание то CityCreateDto, а на просмотр  - CityViewDto

marsel-mkh replied:
Не совсем понятен их смысла в данном проекте,предлогаю обсудить на дэйлике

RinatBeybutov commented on src/main/java/school/faang/user_service/dto/skilloffer/SkillOfferDto.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class SkillOfferDto {
```
то же с названием

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+@Slf4j
+@Data
+@Service
+@RequiredArgsConstructor
```
Давай для класса тоже javadoc Напишем:
Напр 
Сервис для сущности {@link ..}

@author ..

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,60 @@
+        }
+        if (recommendation.getContent() == null) {
+            log.error("Рекомендация = null");
+            throw new DataValidationException("Recommendation context cannot be null");
```
Recommendation **content**

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,60 @@
+        }
+        if (recommendation.getContent().isEmpty()) {
+            log.error("Рекомендация пустая");
+            throw new DataValidationException("Recommendation context cannot be empty");
```
Recommendation **content**

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+     * Также проверяет, что предлагаемые в рекомендации навыки существуют в системе.
+     * @param recommendation - объект DTO с рекомендацией
+     */
+    void validateRecommendation(RecommendationDto recommendation) {
```
приватный метод

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+                .map(SkillOfferDto::getSkillId)
+                .toList();
+        for (Long skillId : skillIds) {
+            skillOfferRepository.findById(skillId).orElseThrow(() -> {
```
посмотри в сторону метода existsById, наверняка его не просто так написали

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+        validateRecommendation(recommendation);
+
+        long receiverId = recommendation.getReceiverId();
+        User user = getUser(receiverId);
```
давай изменим имя переменной c user на receiver

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+     * @param recommendationId - идентификатор рекомендации
+     */
+    public void delete(long recommendationId) {
+        recommendationRepository.findById(recommendationId).orElseThrow(() -> {
```
Можно улучшить: либо использовать метод isExist, либо находить сущность, а потом ее передавать в метод удаления

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+    private List<SkillOffer> SkillOffersToEntity(RecommendationDto recommendation) {
+        List<SkillOffer> allSkillOffer = new ArrayList<>();
+        for (SkillOfferDto skillOfferDto : recommendation.getSkillOffers()) {
+            SkillOffer skillOffer = skillOfferRepository.findById(skillOfferDto.getId()).orElseThrow(() -> {
```
валидация на наличие в базе выполняется в другом месте, а тут можно использовать метод репозитория findAllByIds(Iterable ids). Этот код можно будет переписать в функциональном стиле

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+        List<SkillOffer> allSkillOffer = new ArrayList<>();
+        for (SkillOfferDto skillOfferDto : recommendation.getSkillOffers()) {
+            SkillOffer skillOffer = skillOfferRepository.findById(skillOfferDto.getId()).orElseThrow(() -> {
+                log.error("Ошибка: навык с ID {} не найден", skillOfferDto.getId());
```
текст ошибки не очень понятен: в skillRepository и в skillOfferRepository одинаковая ошибка - навык не найден, а репы разные и везде ищешь по первичному ключу. Может разные тексты ошибок должны быть?

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,240 @@
+
+            if (receiver.getSkills().contains(skill)) {
+                List<UserSkillGuarantee> userSkillGuarantees = skill.getGuarantees();
+                List<User> guarantees = userSkillGuarantees.stream()
```
повнимательнее, лист юзеров, а переменная называется гарантии.

marsel-mkh replied:
Тут контекст не гарантии, а гаранты(юзеры-гаранты) 
![image](https://github.com/user-attachments/assets/7e3ddd83-8bd3-465e-a98e-2318376fc576)

RinatBeybutov replied:
![image](https://github.com/user-attachments/assets/5d285408-29ec-481d-9e5a-32af4ad0f6b6)

Commit:
BJS2-65863-write-recommendation
FIX: create RecommendationCreateDto, RecommendationViewDto, SkillOfferViewDto,SkillOfferCreateDto.
fix problems with anyDto,validation method in Service,

marsel-mkh commented:
@RinatBeybutov что думаешь насчет validateRecommendation и saveSkillsOffer может их стоит вынести в отдельные классы? Сейчас код тяжело тестировать

Commit:
BJS2-65863-write-recommendation
FIX: delete method in Service

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,112 @@
+     * @param recommendation DTO для создания рекомендации
+     * @return созданная рекомендация
+     */
+    public RecommendationViewDto giveRecommendation(@NonNull RecommendationCreateDto recommendation) {
```
давай изменим название метода на addRecommendation или createRecommendation

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,112 @@
+     * @param authorId идентификатор пользователя
+     * @return список рекомендаций
+     */
+    public List<RecommendationViewDto> getAllGivenRecommendations(long authorId) {
```
getAllCreatedRecommendation

RinatBeybutov commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationCreateDto.java:
```diff
@@ -0,0 +1,25 @@
+ */
+@Data
+public class RecommendationCreateDto {
+    private Long id;
```
это что за id ?

marsel-mkh replied:
id рекомендации, я его хотел убрать, но он нужен для метода репозитория в сервисе. Думаю можно попробовать убрать его и чтоб вторым параметром отдельно приходил, как ты показал в своем пет проекте.
![image](https://github.com/user-attachments/assets/d15f65e0-ea41-47ab-b1f7-9a404c086c37)

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+        Pageable pageable = PageRequest.of(0, Integer.MAX_VALUE);
+        List<Recommendation> recommendationPage =
+                recommendationRepository.findAllByReceiverId(receiverId, pageable).toList();
+        return recommendationPage.stream()
```
здесь лучше не создавать промежуточный список
ecommendationRepository.findAllByReceiverId(receiverId, pageable)
.stream()
.map(recommendationMapper::toDto)
                .toList(

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+    public List<RecommendationViewDto> getAllGivenRecommendations(long authorId) {
+        Pageable pageable = PageRequest.of(0, Integer.MAX_VALUE);
+        List<Recommendation> recommendationPage = recommendationRepository
+                .findAllByAuthorId(authorId, pageable).toList();
```
так же как и в предыдущем комментарии

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+    private void validateRecommendation(RecommendationCreateDto recommendation) {
+        LocalDateTime sixMonthsAgo = LocalDateTime.now().minusMonths(6);
+        LocalDateTime recommendationUpdateAt = getUpdateAtCreateRecommendation(recommendation)
+                .orElseThrow(() -> new DataValidationException("Update date is not found"));
```
вот здесь ошибку лучше в метод спрятать

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+        List<Long> skillIds = skillOffers.stream()
+                .map(SkillOfferCreateDto::getSkillId)
+                .toList();
+        for (Long skillId : skillIds) {
```
![image](https://github.com/user-attachments/assets/8e8d9fa7-69f3-41c2-911a-123aa523e9f7)

Вот так перепиши, лаконичнее и чище будет выглядеть

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+            }
+        }
+    }
+    private Optional<LocalDateTime> getUpdateAtCreateRecommendation(RecommendationCreateDto recommendation) {
```
пробел между методами

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+    private Optional<LocalDateTime> getUpdateAtCreateRecommendation(RecommendationCreateDto recommendation) {
+        User recommendationAuthor  = getUser(recommendation.getAuthorId());
+        return recommendationAuthor.getRecommendationsGiven().stream()
+                .filter(rec -> {
```
напиши полное название, мне rec ни о чем не говорит

RinatBeybutov commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,281 @@
+                .filter(rec -> {
+                    User receiver = rec.getReceiver();
+                    User createRecommendationReceiver = getUser(receiver.getId());
+                    return receiver.equals(createRecommendationReceiver);
```
как здесь equals работает?

Commit:
BJS2-65863-write-recommendation
FIX: in methods getAllUserRecommendations and getAllGivenRecommendations add new param Pageable and return Page
guarantee -> guarantors
add param id to create and update method
add RecommendationValidator class
add SkillOfferService with method saveSkillsOffer
add Spring validation in DTO
add new method in RecommendationService "getRecommendationEntity"
add new Exception EntityNotFoundException

Commit:
BJS2-65863-write-recommendation
FIX: переработал стрим в методе getUpdateAtCreateRecommendation
спрятал ошибку в getUpdateAtCreateRecommendation
переработал стрим validateSkillOffers

gulnaz21 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(target = "author", ignore = true)
+    @Mapping(target = "receiver", ignore = true)
+    @Mapping(source = "skillOffers", target = "skillOffers")
+    Recommendation ViewDtoToEntity(RecommendationViewDto recommendation);
```
Название метода должно начинаться с маленькой буквы

gulnaz21 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(source = "author.id", target = "authorId")
+    @Mapping(source = "receiver.id", target = "receiverId")
+    @Mapping(source = "skillOffers", target = "skillOffers")
+    RecommendationViewDto toDto(Recommendation recommendation);
```
Название метода toDto можно переименовать в toViewDto, чтобы было ясно, что метод конвертирует объект в RecommendationViewDto

gulnaz21 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(target = "author", ignore = true)
+    @Mapping(target = "receiver", ignore = true)
+    @Mapping(source = "skillOffers", target = "skillOffers")
+    Recommendation CreateDtoToEntity(RecommendationCreateDto recommendation);
```
название метода с маленькой буквы

gulnaz21 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    @Mapping(source = "author.id", target = "authorId")
+    @Mapping(source = "receiver.id", target = "receiverId")
+    @Mapping(source = "skillOffers", target = "skillOffers")
+    RecommendationCreateDto toCreateDto(Recommendation recommendation);
```
Учитывая, что маппер работает с двумя различными типами DTO, возможно, имеет смысл разделить его на два отдельных маппера. Это повысит читаемость и упростит поддержку кода в будущем

RinatBeybutov replied:
по поводу разделения я не согласен, этот маппер связан с рекомендациями, лучше не плодить большое количество классов

gulnaz21 commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    //This mapper for work with ViewDto
+    @Mapping(target = "author", ignore = true)
+    @Mapping(target = "receiver", ignore = true)
+    @Mapping(source = "skillOffers", target = "skillOffers")
```
Думаю, эта строчка лишняя, так как MapStruct автоматически сопоставляет свойства с одинаковыми именами

RinatBeybutov replied:
это замечание оставлено без изменений

gulnaz21 commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,201 @@
+     * @return List<SkillOffer> - список предложенных навыков
+     * @throws DataValidationException если навык не найден
+     */
+    private List<SkillOffer> SkillOffersToEntity(RecommendationCreateDto recommendation) {
```
Название метода должно начинаться с маленькой буквы

gulnaz21 commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    private User getUser(long userId) {
```
У тебя уже был подобный метод в RecommendationService

gulnaz21 commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Transactional
+    public void saveSkillsOffer(@NonNull RecommendationCreateDto recommendation, long recommendationId) {
+        if (recommendation.getSkillOffers() == null || recommendation.getSkillOffers().isEmpty()) {
```
можно вынести recommendation.getSkillOffers() в отдельную переменную

gulnaz21 commented on src/main/java/school/faang/user_service/validation/RecommendationValidator.java:
```diff
@@ -0,0 +1,131 @@
+        LocalDateTime sixMonthsAgo = LocalDateTime.now().minusMonths(6);
+        LocalDateTime lastRecommendationDate = getUpdateAtCreateRecommendation(recommendation);
+
+        if (lastRecommendationDate.isBefore(sixMonthsAgo)) {
```
Здесь стоит использовать метод isAfter(), так как в текущей реализации ты проверяешь была ли сделана  последняя рекомендация более чем шесть месяцев назад, данный метод возвращает true и выкидывает ошибку, но при таком варианте метод должен работать корректно, тогда как в случае выбрасывания ошибки, нам нужно чтобы данная дата рекомендации была сделана позже, чем 6 месяцев. 
Поэтому давай заменим на  if (lastRecommendationDate.isAfter(sixMonthsAgo))

Commit:
BJS2-65863-write-recommendation
FIX: поправил всякие мелочи(пробелы,названия)

Commit:
BJS2-65863-write-recommendation
FIX: "Skill offer is Empty" in validateSkillOffers

RinatBeybutov commented on src/main/java/school/faang/user_service/validation/RecommendationValidator.java:
```diff
@@ -0,0 +1,129 @@
+        validateRecommendationTimeInterval(recommendation);
+        validateSkillOffers(recommendation);
+        validateAuthorAndReceiver(recommendation);
+
```
лишний пробел

RinatBeybutov commented on src/main/java/school/faang/user_service/validation/RecommendationValidator.java:
```diff
@@ -0,0 +1,129 @@
+     * @param recommendation - объект DTO с рекомендацией
+     * @return Optional<LocalDateTime> - дата последнего обновления рекомендации
+     */
+    private LocalDateTime getUpdateAtCreateRecommendation(RecommendationCreateDto recommendation) {
```
кмк лучше getRecommendationLastUpdateDate

Commit:
BJS2-6679-test-write-recommendation
убрал пробел и переименовал getRecommendationLastUpdateDate

RinatBeybutov commented on src/main/java/school/faang/user_service/mapper/RecommendationMapper.java:
```diff
@@ -0,0 +1,32 @@
+    //This mapper for work with ViewDto
+    @Mapping(target = "author", ignore = true)
+    @Mapping(target = "receiver", ignore = true)
+    @Mapping(source = "skillOffers", target = "skillOffers")
```
у тебя методы viewDtoToEntity и toCreateDto вообще нигде не используются, лучше удалить их

Commit:
BJS2-65863-write-recommendation
убрал 2 метода с маппера

Commit:
BJS2-65863-write-recommendation добавил аннотацию валид,
